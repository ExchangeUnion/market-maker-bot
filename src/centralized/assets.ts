import { BigNumber } from 'bignumber.js';
import { Balances, Exchange } from 'ccxt';
import { curry } from 'ramda';
import { interval, Observable } from 'rxjs';
import { map, mapTo, repeatWhen, startWith, take, tap } from 'rxjs/operators';
import { Config } from '../config';
import { Logger } from '../logger';
import { ExchangeAssetAllocation } from '../trade/info';

const logAssetAllocation = (
  logger: Logger,
  source: Observable<ExchangeAssetAllocation>
) => {
  return source.pipe(
    tap(({ baseAssetBalance, quoteAssetBalance }) => {
      logger.info(
        `Base asset balance ${baseAssetBalance.toString()} and quote asset balance ${quoteAssetBalance.toString()}`
      );
    })
  );
};

type GetCentralizedExchangeAssetsParams = {
  config: Config;
  logger: Logger;
  CEX: Exchange;
  CEXfetchBalance$: (exchange: Exchange) => Observable<Balances>;
  convertBalances: (
    config: Config,
    balances: Balances
  ) => ExchangeAssetAllocation;
  logAssetAllocation: (
    logger: Logger,
    source: Observable<ExchangeAssetAllocation>
  ) => Observable<ExchangeAssetAllocation>;
};

const getCentralizedExchangeAssets$ = ({
  config,
  logger,
  CEX,
  CEXfetchBalance$,
  convertBalances,
  logAssetAllocation,
}: GetCentralizedExchangeAssetsParams): Observable<ExchangeAssetAllocation> => {
  const logAssetAllocationWithLogger = curry(logAssetAllocation)(logger);
  if (!config.TEST_MODE) {
    const convertBalancesWithConfig = curry(convertBalances)(config);
    return CEXfetchBalance$(CEX).pipe(
      map(convertBalancesWithConfig),
      logAssetAllocationWithLogger,
      take(1),
      // refetch assets every 30 seconds
      repeatWhen(() => {
        return interval(30000);
      })
    );
  } else {
    const testCentralizedBalances = {
      baseAssetBalance: new BigNumber(
        config.TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE
      ),
      quoteAssetBalance: new BigNumber(
        config.TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE
      ),
    };
    return interval(30000).pipe(
      startWith(testCentralizedBalances),
      mapTo(testCentralizedBalances),
      logAssetAllocationWithLogger
    );
  }
};

export {
  getCentralizedExchangeAssets$,
  GetCentralizedExchangeAssetsParams,
  logAssetAllocation,
};
