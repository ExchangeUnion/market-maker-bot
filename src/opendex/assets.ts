import { combineLatest, interval, Observable } from 'rxjs';
import { map, mergeMap, repeatWhen, take, tap } from 'rxjs/operators';
import { Config } from '../config';
import { Logger } from '../logger';
import { XudClient } from '../proto/xudrpc_grpc_pb';
import { GetBalanceResponse, TradingLimitsResponse } from '../proto/xudrpc_pb';
import { OpenDEXassetAllocation } from '../trade/info';
import {
  LogAssetBalanceParams,
  ParseOpenDEXassetsParams,
} from './assets-utils';

type GetOpenDEXassetsParams = {
  config: Config;
  logger: Logger;
  logBalance: ({ logger, assets }: LogAssetBalanceParams) => void;
  xudClient$: (config: Config) => Observable<XudClient>;
  xudBalance$: (client: XudClient) => Observable<GetBalanceResponse>;
  xudTradingLimits$: (client: XudClient) => Observable<TradingLimitsResponse>;
  parseOpenDEXassets: ({
    balanceResponse,
    tradingLimitsResponse,
    quoteAsset,
    baseAsset,
  }: ParseOpenDEXassetsParams) => OpenDEXassetAllocation;
};

const getOpenDEXassets$ = ({
  config,
  logger,
  logBalance,
  xudClient$,
  xudBalance$,
  xudTradingLimits$,
  parseOpenDEXassets,
}: GetOpenDEXassetsParams): Observable<OpenDEXassetAllocation> => {
  return xudClient$(config).pipe(
    mergeMap(client => {
      return combineLatest(xudBalance$(client), xudTradingLimits$(client));
    }),
    map(([balanceResponse, tradingLimitsResponse]) => {
      return parseOpenDEXassets({
        balanceResponse,
        tradingLimitsResponse,
        quoteAsset: config.QUOTEASSET,
        baseAsset: config.BASEASSET,
      });
    }),
    tap(assets => logBalance({ assets, logger })),
    take(1),
    // refresh assets every 30 seconds
    repeatWhen(() => interval(30000))
  );
};

export { getOpenDEXassets$ };
