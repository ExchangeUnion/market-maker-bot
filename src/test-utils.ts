import { Config } from './config';
import { Level, Loggers } from './logger';

const testConfig = (): Config => {
  return {
    LOG_LEVEL: Level.Trace,
    CEX: 'Binance',
    CEX_API_KEY: '123',
    CEX_API_SECRET: 'abc',
    DATA_DIR: '',
    OPENDEX_CERT_PATH: `${__dirname}/../mock-data/tls.cert`,
    OPENDEX_RPC_HOST: 'localhost',
    OPENDEX_RPC_PORT: '1234',
    MARGIN: '0.06',
    BASEASSET: 'ETH',
    QUOTEASSET: 'BTC',
    CEX_BASEASSET: 'ETH',
    CEX_QUOTEASSET: 'BTC',
    TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE: '321',
    TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE: '123',
    TEST_MODE: true,
  };
};

const getLoggers = (): Loggers => {
  const mockLogger = {
    warn: () => {},
    info: () => {},
    verbose: () => {},
    debug: () => {},
    trace: () => {},
    error: () => {},
  };
  return ({
    global: mockLogger,
    centralized: mockLogger,
    opendex: mockLogger,
  } as unknown) as Loggers;
};

type TestError = {
  code: string | number;
  message: string;
};

export { getLoggers, testConfig, TestError };
