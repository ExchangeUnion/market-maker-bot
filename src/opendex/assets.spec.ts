import { Observable } from 'rxjs';
import { TestScheduler } from 'rxjs/testing';
import { XudClient } from '../proto/xudrpc_grpc_pb';
import { GetBalanceResponse, TradingLimitsResponse } from '../proto/xudrpc_pb';
import { getLoggers, testConfig } from '../test-utils';
import { getOpenDEXassets$ } from './assets';

type OpenDEXassetsInputEvents = {
  xudClient$: string;
  xudBalance$: string;
  xudTradingLimits$: string;
  unsubscribe: string;
};

let testScheduler: TestScheduler;

const assertOpenDEXassets = (
  inputEvents: OpenDEXassetsInputEvents,
  expected: string
) => {
  testScheduler.run(helpers => {
    const { cold, expectObservable } = helpers;
    const parseOpenDEXassets = ({ balanceResponse }: any) => balanceResponse;
    const getXudBalance$ = () => {
      return (cold(inputEvents.xudBalance$) as unknown) as Observable<
        GetBalanceResponse
      >;
    };
    const getXudTradingLimits$ = () => {
      return (cold(inputEvents.xudTradingLimits$) as unknown) as Observable<
        TradingLimitsResponse
      >;
    };
    const getXudClient$ = () => {
      return (cold(inputEvents.xudClient$) as unknown) as Observable<XudClient>;
    };
    const logBalance = () => {};
    const openDEXassets$ = getOpenDEXassets$({
      logBalance,
      config: testConfig(),
      logger: getLoggers().global,
      xudClient$: getXudClient$,
      xudBalance$: getXudBalance$,
      xudTradingLimits$: getXudTradingLimits$,
      parseOpenDEXassets,
    });
    expectObservable(openDEXassets$, inputEvents.unsubscribe).toBe(expected);
  });
};

describe('getOpenDEXassets$', () => {
  beforeEach(() => {
    testScheduler = new TestScheduler((actual, expected) => {
      expect(actual).toEqual(expected);
    });
  });

  test('waits for balance and tradinglimits before emitting', () => {
    const inputEvents = {
      xudBalance$: '1s a',
      xudClient$: '1s a',
      xudTradingLimits$: '1s a',
      unsubscribe: '93s !',
    };
    const expected = '2s a 31999ms a 29999ms a';
    assertOpenDEXassets(inputEvents, expected);
  });
});
