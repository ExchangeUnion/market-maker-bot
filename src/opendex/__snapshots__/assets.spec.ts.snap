// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenDEX xudBalanceToExchangeAssetAllocation error baseAssetBalance 1`] = `"OpenDEX balance does not include balance for base asset ETH or quote asset BTC."`;
